### PR TITLE
[BEAM-793] Add backoff support in JdbcIO Write

### PR DESCRIPTION
**Pull request by jbonofre**
Follow this checklist to help us incorporate your contribution quickly and easily:

 - [X] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [X] Each commit in the pull request should have a meaningful subject line and body.
 - [X] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [X] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [X] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [X] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

### CODE REVIEW

**Comment by jbonofre** [comment]
retest this please

File: sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java [hunk 1]
```diff
@@ -547,6 +559,16 @@ public void teardown() throws Exception {
       return toBuilder().setBatchSize(batchSize).build();
     }
 
+    /**
+     * Enable backoff and retry support to write.
```
Comment (jkff) [inline sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java]: Is there ever a reason not to enable it?

File: sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java [hunk 2]
```diff
@@ -547,6 +559,16 @@ public void teardown() throws Exception {
...
+     *
+     * @param backOffEnabled True to enable backoff support, false else.
```
Comment (jkff) [inline sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java]: These parameter and return value javadocs are not useful.

File: sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java [hunk 3]
```diff
@@ -589,9 +617,8 @@ public void startBundle() {
       public void processElement(ProcessContext context) throws Exception {
         T record = context.element();
 
-        preparedStatement.clearParameters();
-        spec.getPreparedStatementSetter().setParameters(record, preparedStatement);
-        preparedStatement.addBatch();
+        records.add(record);
+        processRecord(record);
```
Comment (jkff) [inline sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java]: Any reason why `processRecord` is needed here? Can we just do it in one place, in `executeBatchWithBackoff`?

File: sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java [hunk 4]
```diff
@@ -600,16 +627,56 @@ public void processElement(ProcessContext context) throws Exception {
         }
       }
 
+      public void processRecord(T record) throws RuntimeException {
```
Comment (jkff) [inline sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java]: `throws RuntimeException` is redundant and doesn't need to be specified.

File: sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java [hunk 5]
```diff
@@ -600,16 +627,56 @@ public void processElement(ProcessContext context) throws Exception {
...
+          BackOff backoff = BUNDLE_WRITE_BACKOFF.backoff();
+          while (true) {
+            // Batch upsert entities.
+            try {
+              int[] updates = preparedStatement.executeBatch();
+              connection.commit();
+              // Break if the commit threw no exception.
+              break;
+            } catch (SQLException exception) {
+              if (!BackOffUtils.next(sleeper, backoff)) {
```
Comment (jkff) [inline sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java]: There should be some logging here

File: sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java [hunk 6]
```diff
@@ -600,16 +627,56 @@ public void processElement(ProcessContext context) throws Exception {
...
+          Sleeper sleeper = Sleeper.DEFAULT;
+          BackOff backoff = BUNDLE_WRITE_BACKOFF.backoff();
+          while (true) {
+            // Batch upsert entities.
+            try {
+              int[] updates = preparedStatement.executeBatch();
+              connection.commit();
+              // Break if the commit threw no exception.
+              break;
+            } catch (SQLException exception) {
```
Comment (jkff) [inline sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java]: Is the connection and the `preparedStatement` guaranteed to be in a good (reusable) state after any possible `SQLException`, either in `executeBatch` or in `commit`? Is `preparedStatement` guaranteed to be cleared?

**Comment by jbonofre** [comment]
Rebase and use backoff by default. @jkff what kind of test do you have in mind ? Thanks !

**Comment by jbonofre** [comment]
retest this please

**Comment by jbonofre** [comment]
retest this please

File: sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java [hunk 7]
```diff
@@ -563,10 +574,15 @@ public PDone expand(PCollection<T> input) {
...
+      private static final int MAX_RETRIES = 5;
+      private static final FluentBackoff BUNDLE_WRITE_BACKOFF =
+              FluentBackoff.DEFAULT
+                      .withMaxRetries(MAX_RETRIES).withInitialBackoff(Duration.standardSeconds(5));
+
       private DataSource dataSource;
       private Connection connection;
       private PreparedStatement preparedStatement;
-      private int batchCount;
+      private List<T> records = new ArrayList();
```
Comment (jkff) [inline sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java]: This uses a raw type. Should be `new ArrayList<>()`

File: sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java [hunk 8]
```diff
@@ -582,34 +598,59 @@ public void setup() throws Exception {
 
       @StartBundle
       public void startBundle() {
-        batchCount = 0;
+        // nothing to do
```
Comment (jkff) [inline sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java]: Remove the method then?

File: sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java [hunk 9]
```diff
@@ -582,34 +598,59 @@ public void setup() throws Exception {
...
-        batchCount++;
+        records.add(record);
 
-        if (batchCount >= spec.getBatchSize()) {
+        if (records.size() >= spec.getBatchSize()) {
           executeBatch();
         }
       }
 
+      public void processRecord(T record) {
```
Comment (jkff) [inline sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java]: Can be private

File: sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java [hunk 10]
```diff
@@ -582,34 +598,59 @@ public void setup() throws Exception {
...
         executeBatch();
       }
 
-      private void executeBatch() throws SQLException {
-        if (batchCount > 0) {
-          preparedStatement.executeBatch();
-          connection.commit();
-          batchCount = 0;
+      private void executeBatch() throws SQLException, IOException, InterruptedException {
+        if (records.size() > 0) {
```
Comment (jkff) [inline sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java]: Flip condition to reduce nesting

File: sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java [hunk 11]
```diff
@@ -582,34 +598,59 @@ public void setup() throws Exception {
...
+          while (true) {
+            try {
+              // add each record in the statement batch
+              records.stream().forEach(this::processRecord);
+              // execute the batch
+              preparedStatement.executeBatch();
+              // commit the changes
+              connection.commit();
+              break;
+            } catch (SQLException exception) {
```
Comment (jkff) [inline sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java]: Please add logging on errors.

Comment (jkff) [inline sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java]: Is it possible to detect whether this is a deadlock? Does `SQLException` contain such information in a portable way?

**Comment by jbonofre** [comment]
Correct, the current test uses the "happy path". I will add a test to simulate a `SQLException`, going into the backoff path.

File: sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java (hunk 11 above)
Comment (jbonofre) [reply sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java]: Unfortunately, it would depend to vendor error. There's no generic way to detect the `SQLException` is related to a deadlock. We would need to add parsing the `SQLException` depending of the vendor (for instance, detect a `ORA-00060` for Oracle, a `1205` for SQLServer).

**Comment by jbonofre** [comment]
I added a test for the write with backoff: I define a lock table strategy. Then I'm starting a first insert transaction on a table. Before committing this transaction, I'm starting the pipeline which tries to insert in the same table. It results with a lock timeout. We can see the backoff used in the log (retrying). Then, in another thread, I commit the first transaction, and we can see the pipeline resumes and insert.

**Comment by jbonofre** [comment]
Run Java Gradle PreCommit

File: sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java [hunk 12]
```diff
@@ -580,37 +599,59 @@ public void setup() throws Exception {
...
+      private void executeBatch() throws SQLException, IOException, InterruptedException {
+        if (records.size() == 0) {
+          return;
+        }
+        Sleeper sleeper = Sleeper.DEFAULT;
+        BackOff backoff = BUNDLE_WRITE_BACKOFF.backoff();
+        while (true) {
+          try {
+            // add each record in the statement batch
+            records.stream().forEach(this::processRecord);
```
Comment (jkff) [inline sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java]: Hmm now that I think of it, `preparedStatement` can simply be a local variable inside this loop. That will simplify the bookkeeping.

File: sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java [hunk 13]
```diff
@@ -580,37 +599,59 @@ public void setup() throws Exception {
...
+            // execute the batch
+            preparedStatement.executeBatch();
+            // commit the changes
+            connection.commit();
+            break;
+          } catch (SQLException exception) {
```
Comment (jkff) [inline sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java]: It seems that all popular databases report deadlocks the same way - via an SQLState of `"40001"`.

https://db.apache.org/derby/docs/10.0/manuals/develop/develop75.html
https://docs.oracle.com/javadb/10.8.3.0/devguide/cdevconcepts32861.html
https://dev.mysql.com/doc/refman/5.7/en/innodb-deadlock-example.html
https://support.microsoft.com/en-us/help/963957/40001-microsoft-odbc-sql-server-driver-sql-server-your-transaction-was
https://stackoverflow.com/questions/26383624/postgres-exceptions-and-java

and so on.

File: sdks/java/io/jdbc/src/test/java/org/apache/beam/sdk/io/jdbc/JdbcIOTest.java [hunk 14]
```diff
@@ -207,40 +208,40 @@ private static void addInitialData(DataSource dataSource, String tableName)
...
     PCollection<TestRow> rows =
-        pipeline.apply(
-            JdbcIO.<TestRow>read()
-                .withDataSourceConfiguration(JdbcIO.DataSourceConfiguration.create(dataSource))
-                .withQuery(String.format("select name,id from %s where name = ?", readTableName))
-                .withStatementPreparator(
-                    (preparedStatement) -> preparedStatement.setString(1, getNameForSeed(1)))
-                .withRowMapper(new JdbcTestHelper.CreateTestRowOfNameAndId())
-                .withCoder(SerializableCoder.of(TestRow.class)));
+      pipeline.apply(
```
Comment (jkff) [inline sdks/java/io/jdbc/src/test/java/org/apache/beam/sdk/io/jdbc/JdbcIOTest.java]: Please revert the unrelated formatting changes if easy.

File: sdks/java/io/jdbc/src/test/java/org/apache/beam/sdk/io/jdbc/JdbcIOTest.java [hunk 15]
```diff
@@ -288,6 +289,71 @@ public void testWrite() throws Exception {
...
+            + tableName)) {
+          resultSet.next();
+          int count = resultSet.getInt(1);
+          // here we have the record inserted by the first transaction (by hand), and a second one
+          // inserted by the pipeline
+          Assert.assertEquals(2, count);
+        }
+      }
+    } finally {
+      DatabaseTestHelper.deleteTable(dataSource, tableName);
```
Comment (jkff) [inline sdks/java/io/jdbc/src/test/java/org/apache/beam/sdk/io/jdbc/JdbcIOTest.java]: This should be in a `Teardown` method - we can fail at various points above and then the table won't be deleted.

File: sdks/java/io/jdbc/src/test/java/org/apache/beam/sdk/io/jdbc/JdbcIOTest.java [hunk 16]
```diff
@@ -288,6 +289,71 @@ public void testWrite() throws Exception {
...
+                        "org.apache.derby.jdbc.ClientDriver",
+                        "jdbc:derby://localhost:" + port + "/target/beam"))
+                .withStatement(String.format("insert into %s values(?, ?)", tableName))
+                .withPreparedStatementSetter(
+                    (element, statement) -> {
+                      statement.setInt(1, element.getKey());
+                      statement.setString(2, element.getValue());
+                    }));
+
+    // starting a thread to perform the commit later, while the pipeline is running into the backoff
```
Comment (jkff) [inline sdks/java/io/jdbc/src/test/java/org/apache/beam/sdk/io/jdbc/JdbcIOTest.java]: Is there a way to verify that there is actually some backoff happening? E.g. use a `@Rule ExpectedLogs` (from `org.apache.beam.sdk.testing`)

File: sdks/java/io/jdbc/src/test/java/org/apache/beam/sdk/io/jdbc/JdbcIOTest.java [hunk 17]
```diff
@@ -79,6 +79,7 @@ public static void startDatabase() throws Exception {
 
     LOG.info("Starting Derby database on {}", port);
 
+    System.setProperty("derby.locks.waitTimeout", "2");
```
Comment (jkff) [inline sdks/java/io/jdbc/src/test/java/org/apache/beam/sdk/io/jdbc/JdbcIOTest.java]: Add a comment why?

File: sdks/java/io/jdbc/src/test/java/org/apache/beam/sdk/io/jdbc/JdbcIOTest.java (hunk 15 above)
Comment (jbonofre) [reply sdks/java/io/jdbc/src/test/java/org/apache/beam/sdk/io/jdbc/JdbcIOTest.java]: As each test use a different table, and we use a derby local storage, that's not a big deal IMHO. However I move this in a `@After` method.

**Comment by jbonofre** [comment]
@jkff I updated according to your comment: detect deadlock based on the SQLState (but configurable by user to adapt to any database provider).

File: sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java [hunk 18]
```diff
@@ -547,6 +564,19 @@ public void teardown() throws Exception {
       return toBuilder().setBatchSize(batchSize).build();
     }
 
+    /**
+     * If a SQL exception occurs, {@link Write} is able to detect it's a deadlock
+     * based on the exception SQL state (see {@link SQLException#getSQLState()}).
+     * In that case, {@link Write} will use a backoff policy to retry.
+     *
+     * @param lockSqlState the {@link SQLException#getSQLState()} identifying a deadlock.
+     * @return the corresponding {@link Write}.
```
Comment (jkff) [inline sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java]: This `@return` javadoc provides no information and should be removed.

File: sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java [hunk 19]
```diff
@@ -577,55 +611,78 @@ public void setup() throws Exception {
...
+            for (T record : records) {
+              processRecord(record, preparedStatement);
+            }
+            // execute the batch
+            preparedStatement.executeBatch();
+            // commit the changes
+            connection.commit();
+            break;
+          } catch (SQLException exception) {
+            if (!exception.getSQLState().equals(spec.getLockSqlState())) {
```
Comment (jkff) [inline sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java]: Hmm, a `.equals()` here seems very restrictive. What if the database has multiple different `SQLState` values corresponding to different kinds of deadlocks? What if its `SQLState` includes both the error code and some kind of arbitrary error message? I think, if you want to customize this, it should be a `java.util.function.Predicate<String>` or perhaps a `Predicate<SQLException>` returning whether to retry or not.

File: sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java [hunk 20]
```diff
@@ -547,6 +564,19 @@ public void teardown() throws Exception {
       return toBuilder().setBatchSize(batchSize).build();
     }
 
+    /**
+     * If a SQL exception occurs, {@link Write} is able to detect it's a deadlock
+     * based on the exception SQL state (see {@link SQLException#getSQLState()}).
+     * In that case, {@link Write} will use a backoff policy to retry.
```
Comment (jkff) [inline sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java]: What's the default value?

File: sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java [hunk 21]
```diff
@@ -577,55 +611,78 @@ public void setup() throws Exception {
...
+            // execute the batch
+            preparedStatement.executeBatch();
+            // commit the changes
+            connection.commit();
+            break;
+          } catch (SQLException exception) {
+            if (!exception.getSQLState().equals(spec.getLockSqlState())) {
+              throw exception;
+            }
+            LOG.warn("Lock detected, retrying", exception);
```
Comment (jkff) [inline sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java]: Lock -> deadlock (these are different concepts)

File: sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java [hunk 22]
```diff
@@ -577,55 +611,78 @@ public void setup() throws Exception {
...
+            // clean up the statement batch and the connection state
+            preparedStatement.clearBatch();
+            connection.rollback();
```
Comment (jkff) [inline sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java]: Is `rollback()` always valid in this case? What if the database auto-rollbacks deadlocked transactions: is it valid to call `rollback()` again even though there is no currently active transaction? Or is there always a default one?

File: sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java [hunk 23]
```diff
@@ -577,55 +611,78 @@ public void setup() throws Exception {
...
+            if (!BackOffUtils.next(sleeper, backoff)) {
+              // we tried the max number of times
+              throw exception;
+            }
+          } finally {
+            preparedStatement.close();
```
Comment (jkff) [inline sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java]: This should be try with resources.

**Comment by jbonofre** [comment]
@jkff I updated the PR according to your comments.

File: sdks/java/io/jdbc/src/test/java/org/apache/beam/sdk/io/jdbc/JdbcIOTest.java [hunk 24]
```diff
@@ -288,6 +298,87 @@ public void testWrite() throws Exception {
...
+    pipeline
+        .apply(Create.of(Collections.singletonList(KV.of(1, "TEST"))))
+        .apply(
+            JdbcIO.<KV<Integer, String>>write()
+                .withDataSourceConfiguration(
+                    JdbcIO.DataSourceConfiguration.create(
+                        "org.apache.derby.jdbc.ClientDriver",
+                        "jdbc:derby://localhost:" + port + "/target/beam"))
+                .withStatement(String.format("insert into %s values(?, ?)", tableName))
+                .withDeadlockPredicate(new JdbcIO.DeadlockPredicate() {
```
Comment (jkff) [inline sdks/java/io/jdbc/src/test/java/org/apache/beam/sdk/io/jdbc/JdbcIOTest.java]: Use a lambda?

File: sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java [hunk 25]
```diff
@@ -506,13 +533,24 @@ public void teardown() throws Exception {
     void setParameters(T element, PreparedStatement preparedStatement) throws Exception;
   }
 
+  /**
+   * An interface used to detect deadlock when a {@link SQLException} occurs.
+   * If {@link DeadlockPredicate#test(SQLException)} returns true, {@link Write} tries
+   * to replay the statements.
+   */
+  @FunctionalInterface
+  public interface DeadlockPredicate extends Serializable {
```
Comment (jkff) [inline sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java]: Maybe call this `RetryStrategy` in case the user wants to retry more kinds of `SQLException`s (though by default we should retry only deadlocks), and rename `test` to `apply`? (accordingly updating documentation)

Comment (jbonofre) [reply sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java]: +1 great idea !

File: sdks/java/io/jdbc/src/test/java/org/apache/beam/sdk/io/jdbc/JdbcIOTest.java (hunk 24 above)
Comment (jbonofre) [reply sdks/java/io/jdbc/src/test/java/org/apache/beam/sdk/io/jdbc/JdbcIOTest.java]: +1

**Comment by jbonofre** [comment]
Updating according to the latest comments. I will merge when Jenkins will be happy ;)

**Comment by jbonofre** [comment]
Run Java Gradle PreCommit

**Comment by jbonofre** [comment]
retest this please

**Comment by jbonofre** [comment]
Run Java Gradle PreCommit

**Comment by jbonofre** [comment]
Run Java Gradle PreCommit

**Comment by jbonofre** [comment]
Run Java Gradle PreCommit

**Comment by jbonofre** [comment]
Run Java Gradle PreCommit

**Comment by jbonofre** [comment]
Run Java Gradle PreCommit

**Comment by jbonofre** [comment]
I'm fixing gradle build.

**Comment by jbonofre** [comment]
gradle build is fixed now. Waiting Jenkins to merge.

**Comment by jbonofre** [comment]
Run Java Gradle PreCommit

**Comment by jbonofre** [comment]
Run Java PreCommit

**Comment by jbonofre** [comment]
Run Java PreCommit
